sdks/python/apache_beam/examples/avro_nyc_trips.py

```diff
@@ -0,0 +1,282 @@
+        {
+            'name': 'service', 'type': 'string'
+        },
+        {
+            'name': 'day', 'type': 'string'
+        },
+        {
+            'name': 'total_price', 'type': 'double'
+        },
+        {
+            'name': 'total_driver_pay', 'type': 'double'
+        },
+        {
+            'name': 'total_trip_miles', 'type': 'double'
+        },
+        {
+            'name': 'total_trip_minutes', 'type': 'double'
+        },
+        {
+            'name': 'total_number_of_trips', 'type': 'long'
+        },
+        {
+            'name': 'price_per_trip', 'type': 'double'
+        },
+        {
+            'name': 'price_per_mile', 'type': 'double'
+        },
+        {
+            'name': 'price_per_minute', 'type': 'double'
+        },
+        {
+            'name': 'driver_pay_per_trip', 'type': 'double'
+        },
+        {
+            'name': 'driver_pay_per_mile', 'type': 'double'
+        },
+        {
+            'name': 'driver_pay_per_minute', 'type': 'double'
+        },
+        {
+            'name': 'miles_per_hour', 'type': 'double'
+        },
+    ],
+    'name': 'nyc_trip_prices',
+    'type': 'record',
+}
+
+
+class CreateKeyWithServiceAndDay(beam.DoFn):
+  """Creates a key, value group.
+
+  The key is the combination of the HVFHS Code converted to the ride-sharing
+  company name and the day of the week of the ride. The value is the original
+  record dictionary object.
+  """
+  def process(self, record: dict):
+    options = {
+        'HV0002': 'Juno', 'HV0003': 'Uber', 'HV0004': 'Via', 'HV0005': 'Lyft'
+    }
+    service = options.get(record['hvfhs_license_num'])
+    if service:
+      timestamp = None
+      for k in ('request_datetime',
+                'on_scene_datetime',
+                'pickup_datetime',
+                'dropoff_datetime'):
+        timestamp = timestamp or record[k]
+        if timestamp:
+          break
+
+      day_of_the_week = datetime.datetime.fromtimestamp(
+          timestamp / 1000.0,
+          tz=pytz.timezone('America/New_York')).strftime('%a')
+
+      yield (service, day_of_the_week), record
+
+
+class CalculatePricePerAttribute(beam.CombineFn):
+  """Calculates the price for attribute.
+
+  Calculates the total driver pay, price, miles, minutes, and trips per for
+  hire vehicle service. And calculates the price per mile, minute, and trip
+  for both the driver and passenger.
+  """
+  def create_accumulator(self):
+    total_price = 0.0
+    total_driver_pay = 0.0
+    total_trip_miles = 0.0
+    total_trip_time = 0.0
+    total_number_of_trips = 0
+    accumulator = (
+        total_price,
+        total_driver_pay,
+        total_trip_miles,
+        total_trip_time,
+        total_number_of_trips,
+    )
+    return accumulator
+
+  def add_input(self, accumulator, record):
+    (
+        total_price,
+        total_driver_pay,
+        total_trip_miles,
+        total_trip_time,
+        total_number_of_trips,
+    ) = accumulator
```

Comment:
Have you considered using `collections.namedtuple()` instead?

Reply:
I just tried setting the `create_accumulator` function to use `collections.namedtuple()` instead. However, it didn't quite work as I was expecting. Here is what I did:
```python
def create_accumulator(self):
    accumulator = collections.namedtuple(
        'Accumulator',
        (
            'total_price',
            'total_driver_pay',
            'total_trip_miles',
            'total_trip_time',
            'total_number_of_trips',
        ),
        defaults=[0.0, 0.0, 0.0, 0.0, 0],
    )()
    return accumulator
```

But upon accessing the accumulator in the other functions, I was of type `tuple` rather than an instance of the `namedtuple` object.

```python
def add_input(self, accumulator, record):
    # prints <class 'tuple'>
    print(type(accumulator))

    # rather than <class '__main__.Accumulator'>
    # this statement does not work
    print(accumulator.total_price)
```

If you wanted me to use the `collections.namedtuple` in the `create_accumulator` function, but still get the variables from indexing the tuple object in the other functions, I could do that.

Reply:
thanks. Looks like it gets coerced to a tuple. let's leave it as a tuple then.